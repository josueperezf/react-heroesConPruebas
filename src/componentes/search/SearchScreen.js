import React, { useMemo } from 'react';
import { useForm } from '../../hooks/useForm';
import { HeroeCard } from '../heroes/HeroeCard';
import queryString from 'query-string';
import { useNavigate, useLocation  } from 'react-router-dom';
import { getHeroePorName } from '../../selectores/getHeroePorName';

export const SearchScreen = () => {
    // para obtener los parametros url por get
    const location = useLocation()
    const navigate = useNavigate();
    // se usa el plugin query-string que sirve para convertir en objeto todos los parametros que envien por url
    // recibe el query string, ejemplo ?q=abc, resultado seria {q=abc}
    //{ q = '' } hace que si la variable no se la enviaron por url, entonces coloquele de valor comillas vacias para q no quede undefined
    const { q = '' } = queryString.parse(location.search);
    
    const [ formValues, handleInputChange] = useForm({
        buscar:q,
    });
    const {buscar} = formValues;

    // la siguiente linea hay que ponerla en un memo para evitar que cada vez que alguien escriba haga la busqueda,
    // ya que se programo para que busque solo con el boton submit
    // como la variable busqueda cambia de valor cada vez que alguien escribe en el iput, entonces realizar la busqueda, pero ello no es el requerimiento
    // const heroesFiltered = getHeroePorName(buscar);
    const heroesFiltered = useMemo(() => getHeroePorName(q), [q]);
    // no se para que uso el memo si con la siguiente linea hace lo mismo
    // const heroesFiltered = getHeroePorName(q);

    const handleSearch =(e)=> {
        e.preventDefault();
        navigate(`?q=${buscar}`);
    }
    return (
        <div>
            <h1>SearchScreen</h1>
            <hr/>
            <div className='row'>
                <div className='col-5'>
                    <h4>Search Form</h4>
                    <hr/>
                    <form onSubmit={handleSearch} >
                        <input
                            type='text'
                            placeholder='Buscar Heroe'
                            className='form-control'
                            name='buscar'
                            autoComplete="off"
                            autoCapitalize="true"
                            value={buscar}
                            onChange={ handleInputChange }
                        />
                        <button
                            type='submit'
                            className='btn m-1 w-100 btn-outline-primary'
                        >
                            Buscar
                        </button>
                    </form>
                </div>
                <div className='col-7'>
                    <h4>Resultados</h4>
                    <hr/>
                    {
                        (q === '') &&
                        <div className='alert alert-info'>
                            Busqueda de heroe
                        </div>
                    }

                    {
                        (q !== '' && heroesFiltered.length ===0) &&
                        <div className='alert alert-danger'>
                            Heroe no encontrado {q}
                        </div>
                    }

                    {
                        heroesFiltered.map(heroe =>(
                            <HeroeCard key={heroe.id} {...heroe} />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
