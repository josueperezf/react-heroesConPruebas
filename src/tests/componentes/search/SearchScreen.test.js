import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router";
import { SearchScreen } from "../../../componentes/search/SearchScreen";

describe('Pruebas en <SearchScreen/>',()=>{

    // mostrar los valores por default
    test('debe de mostrarse correctamente con los valores por default', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path='/search' component={()=><SearchScreen/> } />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        // tomo un elemento html que debio renderizarse lo hizo
        expect(wrapper.find('.alert-info').exists()).toBe(true);
    });

    // que lo que yo pase por url como parametro lo coloque en 
    test('debe de mostrar a batman, y el input con el valor del querystring', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path='/search' component={()=><SearchScreen/> } />
            </MemoryRouter>
        );
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de mostrar un error si no se encuentra el heroe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman-xxx-x-x-para-q-no-encuentre']}>
                <Route path='/search' component={()=><SearchScreen/> } />
            </MemoryRouter>
        );
        //si no encuentra al heroe, debe mostrar un div con texto y una clase css llamada alert-danger, si esa clase esta es porque no encontro
        expect(wrapper.find('.alert-danger').exists()).toBe(true);
    });

    test('debe de llamar el push del history', () => {
        const history = {
            push:jest.fn()
        };
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman-xxx-x-x-para-q-no-encuentre']}>
                <Route path='/search' component={()=><SearchScreen history={history} /> } />
            </MemoryRouter>
        );
        wrapper.find('input').simulate('change',{
            target:{
                name:'buscar',
                value:'batman'
            }
        });
        // en la siguiente linea simulo el submit del formulario
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
        // verifico si el push fue llamado con el parametro batman
        expect( history.push ).toHaveBeenCalledWith(`?q=batman`);
    });
     
});