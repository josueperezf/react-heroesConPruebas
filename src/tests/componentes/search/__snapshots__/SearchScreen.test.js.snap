// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> debe de mostrar a batman, y el input con el valor del querystring 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoCapitalize="true"
          autoComplete="off"
          className="form-control"
          name="buscar"
          onChange={[Function]}
          placeholder="Buscar Heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 w-100 btn-outline-primary"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card h-100"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/heroe/dc-batman"
                  onClick={[Function]}
                >
                  Mas...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> debe de mostrarse correctamente con los valores por default 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoCapitalize="true"
          autoComplete="off"
          className="form-control"
          name="buscar"
          onChange={[Function]}
          placeholder="Buscar Heroe"
          type="text"
          value=""
        />
        <button
          className="btn m-1 w-100 btn-outline-primary"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Busqueda de heroe
      </div>
    </div>
  </div>
</div>
`;
